nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -718,16 +718,135 @@ impl StringPrototype {
         // 6. If functionalReplace is false, Set replaceValue to ? ToString(replaceValue).
         let replace_string = to_string(agent, replace_value)?;
         // Everything are strings: `"foo".replace("o", "a")` => use rust's replace
+        let result =
+            s.as_str(agent)
+                .replacen(search_string.as_str(agent), replace_string.as_str(agent), 1);
+        Ok(String::from_string(agent, result).into_value())
+    }
+
+    /// ### [22.1.3.20 String.prototype.replaceAll ( searchValue, replaceValue )](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replaceall)
+    fn replace_all(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+
+        let search_value = args.get(0);
+        let replace_value = args.get(1);
+
+        // 2. If searchValue is neither undefined nor null, then
+        if !search_value.is_null() && !search_value.is_undefined() {
+            // a. Let isRegExp be ? IsRegExp(searchValue).
+            let is_reg_exp = false;
+
+            // b. If isRegExp is true, then
+            if is_reg_exp {
+                // i. Let flags be ? Get(searchValue, "flags").
+                // ii. Perform ? RequireObjectCoercible(flags).
+                // iii. If ? ToString(flags) does not contain "g", throw a TypeError exception.
+                todo!();
+            }
+
+            // c. Let replacer be ? GetMethod(searchValue, %Symbol.replace%).
+            let symbol = WellKnownSymbolIndexes::Replace.into();
+            let replacer = get_method(agent, search_value, symbol)?;
+
+            // d. If replacer is not undefined, Return ? Call(replacer, searchValue, « O, replaceValue »).
+            if let Some(replacer) = replacer {
+                return call_function(
+                    agent,
+                    replacer,
+                    search_value,
+                    Some(ArgumentsList(&[o, replace_value])),
+                );
+            }
+        }
+
+        // 3. Let s be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 4. Let searchString be ? ToString(searchValue).
+        let search_string = to_string(agent, search_value)?;
+
+        // 5. Let functionalReplace be IsCallable(replaceValue).
+        if let Some(functional_replace) = is_callable(replace_value) {
+            // 7. Let searchLength be the length of searchString.
+            let search_length = search_string.len(agent);
+
+            // 8. Let advanceBy be max(1, searchLength).
+            let advance_by = max(1, search_length);
+
+            // 9. Let matchPositions be a new empty List.
+            let mut match_positions: Vec<usize> = vec![];
+
+            // 10. Let position be StringIndexOf(s, searchString, 0).
+            let search_str = search_string.as_str(agent).to_owned();
```

Comment:
**nitpick:** Best as I can see, the `to_owned` is not actually needed here at all. `search_str` is not used after the while loop over the positions which also doesn't do any mutations on `agent`, so `search_str` could just be a borrowed string from the heap.

Reply:
Yep. I don't remember why I had to add it, but it's deleted now and it works.